### PR TITLE
Add additionalPrinterColumns to CRD to get more details about clusters

### LINKED ISSUE #539: Add additionalPrinterColumns to CRD to get more details about clusters
Hi Team,

Currently when we do "kubectl get os" the output is as per below, which is not having much information about the clusters like health, number of nodes, phase etc..

NAME                    AGE
my-first-cluster    28d

Expected output is as per below with the latest CRD changes

NAME               HEALTH   NODES   VERSION   PHASE     AGE
my-first-cluster  GREEN      6            2.7.0     RUNNING   28d

Please find the attachment's for the current CRD and Updated CRD.

[opensearch-crd.log](https://github.com/Opster/opensearch-k8s-operator/files/11818849/opensearch-crd.log)
[opensearch-crd-new.log](https://github.com/Opster/opensearch-k8s-operator/files/11818850/opensearch-crd-new.log)

I hope this will help when we want to list out the clusters with the basic information.

Thanks,

### PR DESCRIPTION
**Pull request by prabhakartalari**
Fixes #539

### CODE REVIEW

**Comment by swoehrl-mw** [comment]
Hi @prabhakartalari . Thank you for opening this PR. Unfortunately you cannot simply edit the CRD manually as it is auto-generated by kubebuilder so would be overwritten. As such you will need to configure the printer columns via the type definitions in `api/v1/opensearch_types.go`. The [kubebuilder docs](https://book.kubebuilder.io/reference/generating-crd.html#additional-printer-columns) have a section on it.

**Comment by prabhakartalari** [comment]
Hi @swoehrl-mw. Sure, I will check. Thanks

**Comment by prabhakartalari** [comment]
@swoehrl-mw  I have updated PR with the changes now. Please let me know if any issues. Thanks !

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
Hi @prabhakartalari. Please make sure you follow all points from https://github.com/Opster/opensearch-k8s-operator/blob/main/docs/developing.md#submitting-a-pr and include the generated CRD in the PR (and also copied to the helm chart).

File: opensearch-operator/api/v1/opensearch_types.go [hunk 1]
```diff
@@ -284,12 +284,18 @@ type ClusterStatus struct {
...
+// +kubebuilder:printcolumn:name="age",type="date",JSONPath=".metadata.creationTimestamp"
 type OpenSearchCluster struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   ClusterSpec   `json:"spec,omitempty"`
-	Status ClusterStatus `json:"status,omitempty"`
+	Spec       ClusterSpec                                      `json:"spec,omitempty"`
+	Status     ClusterStatus                                    `json:"status,omitempty"`
+	AssocConfs map[metav1.ObjectSelector]metav1.AssociationConf `json:"-"`
```
[the remark this answers is not in the document]
Comment (prabhakartalari) [reply opensearch-operator/api/v1/opensearch_types.go:298]: @swoehrl-mw I have added the lines 287 to 291 & 298. These lines will be adding the below code in CRD file opensearch.opster.io_opensearchclusters.yaml 

`  - additionalPrinterColumns:
    - jsonPath: .status.health
      name: health
      type: string
    - description: Available nodes
      jsonPath: .status.availableNodes
      name: nodes
      type: integer
    - description: Elasticsearch version
      jsonPath: .status.version
      name: version
      type: string
    - jsonPath: .status.phase
      name: phase
      type: string
    - jsonPath: .metadata.creationTimestamp
      name: age
      type: date`

i have tested this locally with controller-gen but i have not copied generated CRD files. I will go through the doc's and fix this

[the remark this answers is not in the document]
Comment (prabhakartalari) [reply opensearch-operator/api/v1/opensearch_types.go]: I have removed the below line and tested all the process as described at https://github.com/Opster/opensearch-k8s-operator/blob/main/docs/developing.md#submitting-a-pr

AssocConfs map[metav1.ObjectSelector]metav1.AssociationConf `json:"-"`

Now, the things are working as expected.

File: opensearch-operator/api/v1/opensearch_types.go [hunk 2]
```diff
@@ -284,6 +284,11 @@ type ClusterStatus struct {
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:shortName=os;opensearch
 // Es is the Schema for the es API
+// +kubebuilder:printcolumn:name="health",type="string",JSONPath=".status.health"
+// +kubebuilder:printcolumn:name="nodes",type="integer",JSONPath=".status.availableNodes",description="Available nodes"
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:288]: Both of these fields do not exist in the status object. You need to either remove them here or add them and logic to fill them to the operator.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

File: opensearch-operator/api/v1/opensearch_types.go (hunk 2 above)
Comment (prabhakartalari) [reply opensearch-operator/api/v1/opensearch_types.go:288]: Will spend some time and try , able to add these fields in the status object but trying how to add the logic to the operator.

Comment (prabhakartalari) [reply opensearch-operator/api/v1/opensearch_types.go:290]: pushed the changes.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
@prabhakartalari You also need logic in the operator to fill those new status fields, otherwise they will always be empty.

**Comment by prabhakartalari** [comment]
> @prabhakartalari You also need logic in the operator to fill those new status fields, otherwise they will always be empty.

**Comment by prabhakartalari** [comment]
sorry accidentally closed

**Comment by prabhakartalari** [comment]
> > @prabhakartalari You also need logic in the operator to fill those new status fields, otherwise they will always be empty.

I am seeking some help here, could you help me to point to the code where the phase and version is coming. I will refer and try to add the logic. if not will remove those 2 additional fields and submit the changes. Thanks

**Comment by swoehrl-mw** [comment]
Hi @prabhakartalari.
version and phase are coming from two different places:
https://github.com/Opster/opensearch-k8s-operator/blob/0e3cd4b976935696d78ebb344e59e272f189ae28/opensearch-operator/pkg/reconcilers/cluster.go#L118
https://github.com/Opster/opensearch-k8s-operator/blob/0e3cd4b976935696d78ebb344e59e272f189ae28/opensearch-operator/controllers/opensearchController.go#L225

For the new fields (health and availableNodes) it's actually not that easy. Depending on how you define availableNodes this could just be the sum of the configured replicas (then it would be easy in the pkg/reconcilers/cluster.go to calculate and set that), if it should be the number of running nodes then its more complicated as you'd need to basically read all the statefulsets and sum up the reported pod counts (probably also somewhere in reconcilers/cluster.go). As for health, this will either also have to come from querying the statefulset status and calculating an overarching health based on the pod readyness, or will have to query the opensearch api to get a cluster status.

If you feel this is too much for your first PR I'm fine with this PR just adding the other two fields and leaving a feature request ticket open for someone/sometime else for more status fields.

**Comment by prabhakartalari** [comment]
> Hi @prabhakartalari. version and phase are coming from two different places:
> 
> https://github.com/Opster/opensearch-k8s-operator/blob/0e3cd4b976935696d78ebb344e59e272f189ae28/opensearch-operator/pkg/reconcilers/cluster.go#L118
> 
> 
> https://github.com/Opster/opensearch-k8s-operator/blob/0e3cd4b976935696d78ebb344e59e272f189ae28/opensearch-operator/controllers/opensearchController.go#L225
> 
> For the new fields (health and availableNodes) it's actually not that easy. Depending on how you define availableNodes this could just be the sum of the configured replicas (then it would be easy in the pkg/reconcilers/cluster.go to calculate and set that), if it should be the number of running nodes then its more complicated as you'd need to basically read all the statefulsets and sum up the reported pod counts (probably also somewhere in reconcilers/cluster.go). As for health, this will either also have to come from querying the statefulset status and calculating an overarching health based on the pod readyness, or will have to query the opensearch api to get a cluster status.
> 
> If you feel this is too much for your first PR I'm fine with this PR just adding the other two fields and leaving a feature request ticket open for someone/sometime else for more status fields.

Sure, I submitted the PR by removeing the status & availablenodes related code and submitted an https://github.com/Opster/opensearch-k8s-operator/issues/553 issue as a future enhancement request.

File: opensearch-operator/api/v1/opensearch_types.go [hunk 3]
```diff
@@ -284,6 +284,9 @@ type ClusterStatus struct {
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:shortName=os;opensearch
 // Es is the Schema for the es API
+// +kubebuilder:printcolumn:name="version",type="string",JSONPath=".status.version",description="Elasticsearch version"
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:287]: Please change`Elasticsearch` to `Opensearch` as we only handle opensearch with this operator.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
One nitpick, otherwise LGTM

File: opensearch-operator/api/v1/opensearch_types.go (hunk 3 above)
Comment (prabhakartalari) [reply opensearch-operator/api/v1/opensearch_types.go]: done

**Review by swoehrl-mw — APPROVED** [review]

**Pull request merged**